lib/phoenix/endpoint.ex

```diff
@@ -657,68 +654,14 @@ defmodule Phoenix.Endpoint do
 
       @doc false
       def __sockets__, do: unquote(Macro.escape(sockets))
-
-      @doc false
-      def __handler__(%{path_info: path} = conn, opts), do: do_handler(path, conn, opts)
-      unquote(dispatches)
-      defp do_handler(_path, conn, opts), do: {:plug, conn, __MODULE__, opts}
     end
   end
 
-  defp socket_paths(endpoint, path, socket, opts) do
-    paths = []
-    websocket = Keyword.get(opts, :websocket, true)
-    longpoll = Keyword.get(opts, :longpoll, false)
-
-    paths =
-      if websocket do
-        config = Phoenix.Socket.Transport.load_config(websocket, Phoenix.Transports.WebSocket)
-        {conn_ast, match_path} = socket_path(path, config)
-        [{match_path, :websocket, conn_ast, socket, config} | paths]
-      else
-        paths
-      end
-
-    paths =
-      if longpoll do
-        config = Phoenix.Socket.Transport.load_config(longpoll, Phoenix.Transports.LongPoll)
-        plug_init = {endpoint, socket, config}
-        {conn_ast, match_path} = socket_path(path, config)
-        [{match_path, :plug, conn_ast, Phoenix.Transports.LongPoll, plug_init} | paths]
-      else
-        paths
-      end
-
-    paths
-  end
-
-  defp socket_path(path, config) do
-    end_path_fragment = Keyword.fetch!(config, :path)
-
-    {vars, path} =
-      String.split(path <> "/" <> end_path_fragment, "/", trim: true)
-      |> Enum.join("/")
-      |> Plug.Router.Utils.build_path_match()
-
-      conn_ast =
-        if vars == [] do
-          quote do
-            conn
-          end
-        else
-          params =
-            for var <- vars,
-                param = Atom.to_string(var),
-                not match?("_" <> _, param),
-                do: {param, Macro.var(var, nil)}
-
-          quote do
-            params = %{unquote_splicing(params)}
-            %Plug.Conn{conn | path_params: params, params: params}
-          end
-        end
-
-    {conn_ast, path}
+  @doc false
+  defmacro __add_fallback_route__(_) do
+    quote do
+      defp do_match(conn, _, _, _), do: conn
```

Comment:
This is private API and relying on implementation details of Plug.Router, we most likely shouldn't do this.

Reply:
Yep. I was expecting this to raise some red flags. I float a proposal to deal with this above. See https://github.com/phoenixframework/phoenix/pull/5030#discussion_r996490316

Reply:
I dropped a comment here about how to implement it: https://github.com/phoenixframework/phoenix/pull/5030#discussion_r1008718222

TL;DR - convert `__handler__` to a dispatch function and plug it in. :)

Reply:
Done!